File: 71.simplify-path.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class Solution71
{
    /*
    Understadn each conditions which need to clean or move the path
    Leverage the string operations to remove the special cases and make the problem easier
    Leverage stack which can move back

    */
    public string SimplifyPath(string path) {
        Stack<string> stack = new Stack<string>();

        var items = path.Split('/', StringSplitOptions.RemoveEmptyEntries).Where(n => n!= ".");

        foreach(var p in items)
        {
            if(p == "..")
            {
                if(stack.Any())
                {
                    stack.Pop();
                }
            }
            else
            {
                stack.Push(p);
            }
        }

        //Don't forget toe reverse the stack before joining this to the string path.
        return "/"+string.Join('/', stack.Reverse());
    }
}

File: 71_simplify-path.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        if not path or len(path) == 0 : return path

        items = [val for val in path.split('/') if(val and val != '.')]
        # print(items)
        
        routes = []
        for item in items:
            if item == '..':
                if len(routes) > 0:
                    routes.pop()
            else:
                routes.append(item)
        
        return '/'+'/'.join(routes)
    
instance = Solution()
print(instance.simplifyPath('/home/'))
print(instance.simplifyPath('/../'))
print(instance.simplifyPath('/home//foo/'))
